src/me/scana/okgradle/data/JitPackRepository.kt

```diff
@@ -0,0 +1,88 @@
+package me.scana.okgradle.data
+
+import com.google.gson.Gson
+import org.apache.http.HttpStatus
+import org.apache.http.client.HttpClient
+import org.apache.http.client.methods.HttpGet
+import org.apache.http.util.EntityUtils
+import java.net.URLEncoder
+
+class JitPackRepository(private val httpClient: HttpClient, private val gson: Gson) : ArtifactRepository {
+
+    companion object {
+        val JITPACK_URL = "https://jitpack.io/api/search"
+    }
+
+    override fun search(query: String): SearchResult {
+        if (query.isEmpty()) {
+            return SearchResult()
+        }
+
+        return try {
+            var searchResult = artifactIdForName(query)
+            if (shouldRetry(searchResult)) {
+                searchResult = artifactIdForName(query)
+            }
+            searchResult
+        } catch (e: Exception) {
+            SearchResult(error = e)
+        }
+    }
+
+    private fun artifactIdForName(name: String): SearchResult {
+        val args = URLEncoder.encode(name, "UTF-8")
+        val request = HttpGet("$JITPACK_URL?q=$args&limit=10")
+        val response = httpClient.execute(request)
+        val statusLine = response.statusLine
+        val responseEntity = response.entity
+        val responseString = responseEntity?.let { EntityUtils.toString(responseEntity) }
+        EntityUtils.consumeQuietly(responseEntity)
+
+        if (statusLine.statusCode != HttpStatus.SC_OK) {
+            return SearchResult(error = JitPackSearchHttpException(statusLine.statusCode))
+        }
+        return parseSearchResult(responseString)
+    }
+
+    private fun parseSearchResult(responseString: String?): SearchResult {
+        val result = parseJson(responseString)
+        val artifact = parseArtifact(result)
+        return SearchResult(artifact, "")
+    }
+
+    private fun parseJson(responseString: String?): JitPackResult {
+        val artifacts = mutableListOf<Artifact>()
+
+        responseString?.let {
+            val libs = gson.fromJson(it, HashMap<String, List<String>>().javaClass)
```

Comment:
One more tidbit:

With current solution, a new instance of an empty HashMap is created each time we parse a response. It doesn't really matter that much in terms of performance, but we could make it look a little bit nicer :)

This is a most common workaround for this problem with generics:
`val libs = gson.fromJson(it, object : TypeToken<Map<String, List<String>>() {}.type)`

Or we could even do better here, with this (let's put this extension somewhere in this class for now, I can see it get extracted sometime later in the future):
`inline fun <reified T> Gson.fromJson(json: String) = this.fromJson<T>(json, object: TypeToken<T>() {}.type)`

And then you can use it like this:
`val libs: Map<String, List<String>> = gson.fromJson(it)`